### PR TITLE
Ingress resource V0.2

### PR DESCRIPTION
**Pull request by bprashanth**
Eg:

``` yaml
apiVersion: v1
kind: Ingress
metadata:
  name: lbconfig
spec:
  backend:
    serviceName: "foo"
    servicePort: 80
  rules:
  - host: foo.bar.com
    http:
      paths:
      - path: "/foo"
        backend:
          serviceName: "foo"
          servicePort: 80

```

One can create an Ingress with either a default backend, ingress rules, or both but not neither. 
Kubectl get ing:

```
NAME      RULE          BACKEND   ADDRESS
echomap   -             foo:80    0.0.0.0
          foo.bar.com   
          /foo          foo:80
          /bar          bar:8080
          bar.baz.com   
          /baz[.*]      foo:80
          /b?ar         bar:8080
lbmap     -             foo:80    0.0.0.1
          foo.bar.com   
          /foo          foo:80
          /bar          bar:8080
          bar.baz.com   
          /baz[.*]      foo:80
          /b?ar         bar:8080

```

https://github.com/kubernetes/kubernetes/pull/13947  https://github.com/kubernetes/kubernetes/pull/12827 https://github.com/kubernetes/kubernetes/issues/561

### CODE REVIEW

File: pkg/apis/experimental/types.go [hunk 1]
```diff
@@ -508,35 +509,56 @@ type IngressStatus struct {
...
+	// Define the referenced service endpoint which the traffic will be
+	// forwarded to.
 	Backend IngressBackend `json:"backend"`
 }
 
-// IngressBackend describes all endpoints for a given Service, port and protocol.
+// IngressBackend describes all endpoints for a given Service and port.
 type IngressBackend struct {
 	// Specifies the referenced service.
 	ServiceRef api.LocalObjectReference `json:"serviceRef"`
```
Comment (thockin) [inline pkg/apis/experimental/types.go]: LocalObjectReference has one field - Name.  We have mixed conventions on this.  Some places that reference another object in the same namespace say "FooName string" and others say "FooRef LocalObjectreference".

I prefer FooName for terseness.

@bgrant0607 I don't want you do to a full API review, just answer this one point where I argue with myself.  ServiceRef or ServiceName?

File: pkg/apis/experimental/types.go [hunk 2]
```diff
@@ -495,11 +495,12 @@ type IngressList struct {
...
-	// TODO: Add the ability to specify load-balancer IP just like what Service has already done?
-	// A list of rules used to configure the Ingress.
-	// http://<host>:<port>/<path>?<searchpart> -> IngressBackend
-	// Where parts of the url conform to RFC 1738.
-	Rules []IngressRule `json:"rules"`
+	// A default backend capable of servicing requests that don't match any
+	// IngressHost. Currently every IngressSpec requires a default backend.
+	Backend *IngressBackend `json:"backend,omitempty"`
+	// A list of host rules used to configure the Ingress.
+	Hosts []IngressHost `json:"rules"`
```
Comment (thockin) [inline pkg/apis/experimental/types.go]: tag still says rules, as does your PR comment text

File: pkg/apis/experimental/types.go [hunk 3]
```diff
@@ -508,35 +509,56 @@ type IngressStatus struct {
 	LoadBalancer api.LoadBalancerStatus `json:"loadBalancer,omitempty"`
 }
 
-// IngressRule represents the rules mapping the paths under a specified host to the related backend services.
-type IngressRule struct {
+// IngressHost represents the rules mapping the paths under a specified host to
+// the related backend services.
+type IngressHost struct {
 	// Host is the fully qualified domain name of a network host, or its IP
```
Comment (thockin) [inline pkg/apis/experimental/types.go]: What happens if this is an IP and that IP != the claim ?

File: pkg/apis/experimental/types.go [hunk 4]
```diff
@@ -508,35 +509,56 @@ type IngressStatus struct {
...
+	// IngressRule represents a rule to route requests for this IngressHost.
+	IngressRule `json:",inline"`
+}
+
+// IngressRule represents a rule to apply against incoming requests. If the
+// rule is satisfied, the request is routed to the specified backend.
+type IngressRule struct {
+	// Currently mixing different types of rules in a single Ingress is
+	// disallowed, so exactly one of the following must be set.
+	HTTP *HTTPIngressRule `json:"http"`
```
Comment (thockin) [inline pkg/apis/experimental/types.go]: What if you just made this a []HTTPIngressPath.  The YAML seems much nicer

File: pkg/apis/experimental/types.go [hunk 5]
```diff
@@ -508,35 +509,56 @@ type IngressStatus struct {
...
+	// A collection of paths that map requests to IngressBackends.
+	Paths []HTTPIngressPath `json:"paths"`
 }
 
 // IngressPath associates a path regex with an IngressBackend.
 // Incoming urls matching the Path are forwarded to the Backend.
-type IngressPath struct {
-	// Path is a regex matched against the url of an incoming request.
+type HTTPIngressPath struct {
+	// Path is a extended POSIX regex as defined by IEEE Std 1003.1,
```
Comment (thockin) [inline pkg/apis/experimental/types.go]: Full POSIX or the RE2-subset thereof?

File: pkg/apis/experimental/types.go (hunk 1 above)
Comment (karlkfi) [inline pkg/apis/experimental/types.go]: I vote for Ref > Name

File: pkg/apis/experimental/validation/validation.go [hunk 6]
```diff
@@ -333,3 +345,130 @@ func ValidateJobSpecUpdate(oldSpec, spec experimental.JobSpec) errs.ValidationEr
...
+// ValidateIngressName validates that the given name can be used as an Ingress name.
+func ValidateIngressName(name string, prefix bool) (bool, string) {
+	return apivalidation.NameIsDNSSubdomain(name, prefix)
+}
+
+// ValidateIngressSpec tests if required fields in the IngressSpec are set.
+func ValidateIngressSpec(spec *experimental.IngressSpec) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	// TODO: Is a default backend mandatory?
+	if spec.Backend != nil {
```
Comment (thockin) [inline pkg/apis/experimental/validation/validation.go]: What happens if a default backend is not provided but a non-matching path is accessed?  I sort of feel like a default backend is required, not optional

File: pkg/apis/experimental/validation/validation.go [hunk 7]
```diff
@@ -333,3 +345,130 @@ func ValidateJobSpecUpdate(oldSpec, spec experimental.JobSpec) errs.ValidationEr
...
+		allErrs = append(allErrs, errs.NewFieldInvalid("hosts", spec.Hosts, "Either a default backend or a set of host rules are required for ingress."))
+	}
+	if len(spec.Hosts) > 0 {
+		allErrs = append(allErrs, validateIngressHosts(spec.Hosts).Prefix("hosts")...)
+	}
+	return allErrs
+}
+
+// ValidateIngressUpdate tests if required fields in the Ingress are set.
+func ValidateIngressUpdate(oldController, ingress *experimental.Ingress) errs.ValidationErrorList {
```
Comment (thockin) [inline pkg/apis/experimental/validation/validation.go]: oldController is a weird name.

File: pkg/apis/experimental/validation/validation.go [hunk 8]
```diff
@@ -333,3 +345,130 @@ func ValidateJobSpecUpdate(oldSpec, spec experimental.JobSpec) errs.ValidationEr
...
+		return append(allErrs, errs.NewFieldRequired("ingressHosts"))
+	}
+	for _, ih := range ingressHosts {
+		if len(ih.Host) > 0 {
+			// TODO: Disallow ports?
+			host := strings.Split(ih.Host, ":")
+			if valid, errMsg := apivalidation.NameIsDNSSubdomain(host[0], false); !valid {
+				allErrs = append(allErrs, errs.NewFieldInvalid("host", host[0], errMsg))
+			}
+			if len(host) == 2 {
```
Comment (thockin) [inline pkg/apis/experimental/validation/validation.go]: I don't comprehend what a port here represents?  Is this an alternate way to make a claim?  why is this needed?

File: pkg/apis/experimental/validation/validation.go [hunk 9]
```diff
@@ -333,3 +345,130 @@ func ValidateJobSpecUpdate(oldSpec, spec experimental.JobSpec) errs.ValidationEr
...
+			// In addition to this any characters not allowed in a path per
+			// RFC 3986 section-3.3 cannot appear as a literal in the regex.
+			// Consider the example: http://host/valid?#bar, this is a valid
+			// regex that matches valid#bar, which isn't a valid path, because
+			// the path terminates at the first ? or #. A more sophisticated
+			// form of validation would detect that the user is confusing url
+			// regexes with path regexes.
+			_, err := regexp.CompilePOSIX(rule.Path)
+			if err != nil {
+				allErrs = append(allErrs, errs.NewFieldInvalid("path", rule.Path, "httpIngressRule.path needs to be a valid regex."))
```
Comment (thockin) [inline pkg/apis/experimental/validation/validation.go]: s/needs to/must

File: pkg/apis/experimental/validation/validation.go [hunk 10]
```diff
@@ -333,3 +345,130 @@ func ValidateJobSpecUpdate(oldSpec, spec experimental.JobSpec) errs.ValidationEr
...
+}
+
+// validateIngressBackend tests if a given backend is valid.
+func validateIngressBackend(backend *experimental.IngressBackend) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+
+	// All backends must reference a single local service by name, and a single service port by name or number.
+	if len(backend.ServiceRef.Name) == 0 {
+		return append(allErrs, errs.NewFieldRequired("serviceRef.name"))
+	} else if ok, errMsg := apivalidation.ValidateServiceName(backend.ServiceRef.Name, false); !ok {
```
Comment (thockin) [inline pkg/apis/experimental/validation/validation.go]: Something that we don't do, but probably should, that walks the line between validation and REST: cross-object link verification and refcounting.

@lavalamp for ideas, or just to put it on the radar.  We're doing more and more cross-object stuff.  Should we validate that the target of the link exists?  Should we prevent deletion of the linkee as long as the refcount is non-zero?  We'll have to keep back-links.  Complex, but the whole API is getting complex.

File: pkg/apis/experimental/types.go (hunk 1 above)
Comment (bgrant0607) [inline pkg/apis/experimental/types.go]: ServiceName is fine if you don't expect to extend this to refer to services in other namespaces or to other kinds of resources.

File: pkg/apis/experimental/types.go (hunk 5 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: The way i understand this RE2 is close to what perl does (and the normal Compile of golang). I wanted to allow POSIX and let the loadbalancer choose. Haproxy does POSIX. The differences are small enough that I don't *think* it should matter, and using POSIX gives me a nice ieee doc to reference. Correct me and ill change.

File: pkg/apis/experimental/validation/validation.go (hunk 6 above)
Comment (bprashanth) [inline pkg/apis/experimental/validation/validation.go]: I think it's confusing to require a default like gce does. It forces you to take one from your paths and dump it there or create a new one, increasing the size of what should be a tiny resource. I was hoping (and this might be a stretch) to serve some sort of index page on apiserver and default it as the backend. 

In the case you mentioned curl will hang for a bit and then err out the browser will do whatever it does for bad url.

File: pkg/apis/experimental/validation/validation.go (hunk 10 above)
Comment (bprashanth) [inline pkg/apis/experimental/validation/validation.go]: > @lavalamp for ideas, or just to put it on the radar. We're doing more and more cross-object stuff. Should we validate that the target of the link exists?

I want to do this.

File: pkg/apis/experimental/types.go (hunk 1 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: Ref = another object in api (regardless of typeof(ref)), that seems clear to me from a usability perspective. I initially had it as `ServiceName string`, but in theory we could grow a local object ref, or create another object ref to contain more information (like object meta?) and that would be useful here. 

Don't feel strongly eitherway, so defer back to @thockin for final call.

File: pkg/apis/experimental/types.go (hunk 4 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: I wanted to leave the possibility of mixing rules open for now, but can go eitherway on this

File: pkg/apis/experimental/validation/validation.go (hunk 7 above)
Comment (bprashanth) [inline pkg/apis/experimental/validation/validation.go]: old habits die hard

File: pkg/apis/experimental/validation/validation.go (hunk 8 above)
Comment (bprashanth) [inline pkg/apis/experimental/validation/validation.go]: port is allowed in host part of the url according to rfc 1738 (and the uri rfc, which is more recent) so I figured I'd just pass it through and let the loadbalancer controller decide what to do with it. I'm going for least restrictive validation that still produces something that roughly works cross-platform.

File: pkg/apis/experimental/types.go (hunk 3 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: Same as if this is hostname and != hostname on claim. Currently we'll program the loadbalancer with whatever you ask for, validating against the ip in status will only help so much, because you can still screw up hostname right? Though you could argue that we should invalidate what we can. 

Long run an ingress shouldn't be allowed to take another claim hostage. Cross resource validation?

File: pkg/apis/experimental/types.go (hunk 2 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: The name "rules" is less confusing IMO, because that field is really some match rules (host+IngressRule still = Rule). The yaml works out. You could argue that `Rules []IngressHost` is a little confusing, otherwise I'm just going to rename the struct field to Rules. 

IngressHost isn't exposed anywhere and I don't think it's a bad name, but could be convinced of a rename there too. We have enough general general sounding names here, at least IngressHost is specific about the host.

File: pkg/apis/experimental/types.go (hunk 4 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: Hmm, I see what you're saying. Where would ingress specific global information go, into each path (or the claim, but putting loadbalancer configuration into a claim feels messy)? an example of such state is `option forwardfor` which enables insertion of X-Forwarded-For, or `option http-keep-alive` which does what you'd think (allows client server keepalive). 

I actually wanted to go the opposite direction and give the HTTPIngressRule Annotations.

File: pkg/apis/experimental/types.go [hunk 11]
```diff
@@ -464,9 +464,9 @@ type JobCondition struct {
 	Message string `json:"message,omitempty"`
 }
 
-// An Ingress is a way to give services externally-reachable urls. Each Ingress is a
-// collection of rules that allow inbound connections to reach the endpoints defined by
-// a backend.
+// An Ingress is a way to give services externally-reachable urls. Each Ingress
+// is a collection of rules that allow inbound connections to reach the endpoints
+// defined by a backend.
 type Ingress struct {
```
Comment (lavalamp) [inline pkg/apis/experimental/types.go]: Is the purpose of this to help users outside the cluster find servers inside the cluster, or is it to help users inside the cluster find servers outside the cluster?

Comment (lavalamp) [inline pkg/apis/experimental/types.go]: If the former, please make the comment clearer. If the latter, this should be called "Egress".

Comment (karlkfi) [inline pkg/apis/experimental/types.go]: Seems pretty obvious by the comment that it's trying to allow external-to-internal access. Tho if you have subdivision of internal networks along namespaces you might also need it for internal pods to talk to services in another namespaces. Internal-to-internal technically isn't really Egress tho, even if you have to go out through some non-k8s networking or load balancer.

I think the primary use case was for the networking abstraction layer that k8s sits on to create load balancers, network routes, proxies, and/or port forwarding for external-to-internal traffic.

**Comment by karlkfi** [comment]
I'm not clear why we need multiple hosts per ingress. Why not just make a new ingress? The only thing they share is a name, and they could easily share a label instead.

Given that the rules are ordered, you don't need an explicit default service ref. Just put one at the end of the rule list with a nil or "/" path so that it catches everything not caught by previous rules.

It seems like we could flatten and simplify the API significantly to improve UX. How about:

```
type IngressSpec struct {
    Host string // hostname[.domain][:port]
    Rules []IngressRule
}

type IngressRule struct  {
    Path *string // optional. if nil, catches all paths (or isn't http and doesn't have a path)
    Service IngressServiceFacet
}

type IngressServiceFacet struct {
    api.LocalObjectReference // anonymous field for transparent composition
    Port util.IntOrString
}
```

Example usage:

```
apiVersion: v1
kind: Ingress
metadata:
  name: foobar
spec:
  host: foo.bar.com
  rules:
  - path: "/foo"
    service:
      name: "foo"
      port: 80
  - path: "/bar"
    service:
      name: "bar"
      port: 80
  - path: "/"
    service:
      name: "baz"
      port: 80
```

**Comment by bprashanth** [comment]
> I'm not clear why we need multiple hosts per ingress. Why not just make a new ingress? The only thing they share is a name, and they could easily share a label instead.

Atomic updates for all urls associated with a site, blue green deployment type thing for maps.google.com, apis.google.com and so on (some discussion on https://github.com/kubernetes/kubernetes/issues/561). 

> Given that the rules are ordered, you don't need an explicit default service ref. Just put one at the end of the rule list with a nil or "/" path so that it catches everything not caught by previous rules.

I mean yeah, you can say everyone should have a catchall rule, but:
* tcp services shouldn't need that
* you still want a way to say hit this backend, even if all my certs are messed up (we haven't introduced certs yet). With TLS you can't get the "/"

> It seems like we could flatten and simplify the API significantly to improve UX. How about:

Yeah we discussed something like this in https://github.com/kubernetes/kubernetes/pull/13947#issuecomment-141711560, tldr being we want a resource that addresses Ingress, not just HTTP L7 Ingress, so it needs to be usable for defining L4, L7 and any other usecase that comes up that might or might not need paths (hence the IngressRule -> HTTPIngressRule bit). Things like SNI at L4 get a lot harder to think about if you have a flat url based structure. The eventual goal (hopefully eventual == 1.2) is to dovetail the TCP loadbalancer stuff into this. In theory you can have just one resource that has a mix of tcp and http rules, just like a  off the shelf loadbalancer.

**Comment by bprashanth** [comment]
We essentially have your resource, the default backend is optional. The reason we don't have rules going straight to a list of paths is https://github.com/kubernetes/kubernetes/pull/14459#discussion_r40362363 (i.e we might, before this leaves experimental).

**Comment by karlkfi** [comment]
Working through the alternate API proposal above, I got the feeling that the naming doesn't match the structure.

An API resource called "Ingress" sounds like it should be from the perspective of a single Service. It's so much easier to talk about "Ingress routes to a Service", rather than the way the PR proposes, which is a many-to-many relationship or one-to-many relationship from Host(s) to Service Ports. 

I feel like the proposed structure (both mine and the PRs) is more closely describing a set of edge router rules where the router has a Host (like a static IP or dns) and can route/proxy port/paths permutations to different service ports. 

It seems like either the name should change to match the structure (e.g. Route or Router) or the structure should change to match the name (many-to-one Hosts to Service).

**Comment by bprashanth** [comment]
> An API resource called "Ingress" sounds like it should be from the perspective of a single Service. It's so much easier to talk about "Ingress routes to a Service", rather than the way the PR proposes, which is a many-to-many relationship or one-to-many relationship from Host(s) to Service Ports.

Acutally an Ingress is a way to tell the system what to do with an IngressClaim (we don't have this yet). The IngressClaim is a claim on an ip/dns etc, some limited resource that allows entry to the cluster (https://github.com/kubernetes/kubernetes/pull/12827). You're talking specifically about HTTPIngress, which is as you described. Tcp ingress would be straight to service.

The model for httpingress is fanout, if you're going straight to a service the common case is tcp. Http loadbalancer are more expensive in the cloud and people typically want high fanout (from the somewhat biased sampling I've taken of users). At the same time you can't have all your serivces join a single lb for various reasons (some artifical like 1 cert per lb in cloud, some real like saturation or admin policy/classes).

**Comment by karlkfi** [comment]
> Atomic updates for all urls associated with a site, blue green deployment type thing for maps.google.com, apis.google.com and so on (some discussion on #561).

This is achievable if each Ingress is many-to-one Hosts to Service. But you'd be sacrificing the inverse of allowing atomic updates for all services routed by a single host.

> we want a resource that addresses Ingress, not just HTTP L7 Ingress

Hard to do without knowing what fields we want to be able to support in the future... It's definitely reasonable tho to do they whole anonymous field trick on the IngressRule itself, allowing for http and tcp, like we do for probes.

**Comment by karlkfi** [comment]
> The model for httpingress is fanout, if you're going straight to a service the common case is tcp.

If you want tcp routing, just leave off the optional path, no? Why does it need a different model type for that? You just end up with an array of one rule.

**Comment by bprashanth** [comment]
> Hard to do without knowing what fields we want to be able to support in the future... It's definitely reasonable tho to do they whole anonymous field trick on the IngressRule itself, allowing for http and tcp, like we do for probes.

Yeah GLOB was one of the things we discussed in https://github.com/kubernetes/kubernetes/pull/13947#issuecomment-141834479

To add a new type of rule all you have to do today is extend this thing, you don't need to know beforehand:
```
// IngressRule represents a rule to apply against incoming requests. If the
// rule is satisfied, the request is routed to the specified backend.
type IngressRule struct {
	// Currently mixing different types of rules in a single Ingress is
	// disallowed, so exactly one of the following must be set.
	HTTP *HTTPIngressRule `json:"http"`
}
```

> If you want tcp routing, just leave off the optional path, no? Why does it need a different model type for that? You just end up with an array of one rule.

Yeah there are equivalent ways to achieve this wihtout the default backend (https://github.com/kubernetes/kubernetes/pull/13947#issuecomment-141344963), but having a default backend your loadbalancer will hit no matter what else messes up is useful anyway (I mean, you could argue that this is not a valid concept at all for nginx/haproxy/gce l7 etc and that would be a slightly different topic, but all those loadbalancers allow a default_backend setting it for the same reason).

**Comment by bprashanth** [comment]
> Why does it need a different model type for that?

Wait you don't need a new model, this is what your tcp ingress would look like:
```
apiVersion: v1
kind: Ingress
metadata:
  name: no-rules-map
spec:
  backend:
    serviceName: "foo"
    servicePort: 80
  rules: []
```

File: pkg/apis/experimental/types.go (hunk 1 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: We decided on just ServiceName for now, to reduce yaml chatter

File: pkg/apis/experimental/types.go (hunk 2 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: changed Hosts -> Rules

File: pkg/apis/experimental/types.go (hunk 3 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: Ips are no longer allowed, nor are ports

File: pkg/apis/experimental/validation/validation.go (hunk 7 above)
Comment (bprashanth) [inline pkg/apis/experimental/validation/validation.go]: changed to old ingress

File: pkg/apis/experimental/types.go [hunk 12]
```diff
@@ -464,9 +464,10 @@ type JobCondition struct {
 	Message string `json:"message,omitempty"`
 }
 
-// An Ingress is a way to give services externally-reachable urls. Each Ingress is a
-// collection of rules that allow inbound connections to reach the endpoints defined by
-// a backend.
+// Ingress is a collection of rules that allow inbound connections to reach the
+// endpoints defined by a backend. An Ingress can be configured to give services
+// externally-reachable urls, load balance traffic, terminate SSL, offer name
+// based virtual hosting etc.
```
Comment (lavalamp) [inline pkg/apis/experimental/types.go]: Thanks-- I may just be dense but I find this description *much* clearer!

File: pkg/apis/experimental/types.go [hunk 13]
```diff
@@ -495,11 +496,13 @@ type IngressList struct {
...
-	// A list of rules used to configure the Ingress.
-	// http://<host>:<port>/<path>?<searchpart> -> IngressBackend
-	// Where parts of the url conform to RFC 1738.
-	Rules []IngressRule `json:"rules"`
+	// A default backend capable of servicing requests that don't match any
+	// IngressHost. It is optional to allow the loadbalancer controller or
+	// defaulting logic to specify a global default.
+	Backend *IngressBackend `json:"backend,omitempty"`
+	// A list of host rules used to configure the Ingress.
+	Rules []IngressHost `json:"rules"`
```
Comment (thockin) [inline pkg/apis/experimental/types.go]: I know you explained why, but making the type Host and the name Rule bother me,  What if this was `Rules []IngressRule` and what you currently call IngressRule -> IngressRuleValue ?

File: pkg/apis/experimental/types.go [hunk 14]
```diff
@@ -508,35 +511,70 @@ type IngressStatus struct {
...
 
-// IngressRule represents the rules mapping the paths under a specified host to the related backend services.
-type IngressRule struct {
-	// Host is the fully qualified domain name of a network host, or its IP
-	// address as a set of four decimal digit groups separated by ".".
-	// Conforms to RFC 1738.
+// IngressHost represents the rules mapping the paths under a specified host to
+// the related backend services.
+type IngressHost struct {
+	// Host is the fully qualified domain name of a network host, as defined
```
Comment (thockin) [inline pkg/apis/experimental/types.go]: This is a long-winded way of saying it is a DNS_SUBDOMAIN, which we have specced elsewhere.

File: pkg/apis/experimental/types.go [hunk 15]
```diff
@@ -508,35 +511,70 @@ type IngressStatus struct {
...
-	Paths []IngressPath `json:"paths"`
+	// Currently mixing different types of rules in a single Ingress is
+	// disallowed, so exactly one of the following must be set.
+	HTTP *HTTPIngressRule `json:"http"`
+}
+
+// HTTPIngressRule is a list of http selectors pointing to IngressBackends.
+// In the example: http://<host>/<path>?<searchpart> -> IngressBackend where
+// where parts of the url correspond to RFC 3986, this resource will be used
+// to match against everything after the last '/'.
```
Comment (thockin) [inline pkg/apis/experimental/types.go]: Wait, do URL params count here?  Can I spec a path of "/foo?go-get=1" and another as "/foo"?  What if I want to allow URL params to "/bar" but I don't want to spec them here?  "/bar(\?.*)?"  bleeeech

File: pkg/apis/experimental/types.go (hunk 14 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: wait, a string with ip is also a subdomain right?

File: pkg/apis/experimental/types.go (hunk 15 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: That's lb controller specific. We don't validate against it (path should be a valid regex, that's all -- see comment above validation where I call CompilePOSIX). If the other end validates it as a pure path it should surface the error. Otherwise you can totally specify a full url regex in haproxy with the right config options.

File: pkg/apis/experimental/types.go (hunk 14 above)
Comment (thockin) [inline pkg/apis/experimental/types.go]: hmm, technically yes but that is a valid hostname, but no TLD is purely
numeric.  If we're asserting that this is a hostname field and someone
enters an IP, we 're going to do a DNS lookup on it.

File: pkg/apis/experimental/types.go (hunk 15 above)
Comment (thockin) [inline pkg/apis/experimental/types.go]: I'm worried that we're setting people up for bad couplings.  If "/foo?go-get=1" works on some load-balancers and not others, people will get hurt.  Is it worth the energy to restrict this to the path?

File: pkg/apis/experimental/types.go (hunk 14 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: I'm validating against the ip case also, because I think in most cases if someone puts an ip in there they don't expect a dns lookup

File: pkg/apis/experimental/types.go (hunk 15 above)
Comment (bprashanth) [inline pkg/apis/experimental/types.go]: Yeah, ok, will spec it to just path. If someone still wants a urlregex it's their problem.

**Comment by bprashanth** [comment]
```
07:35:12 NAME                    NETWORK             SRC_RANGES RULES   SRC_TAGS TARGET_TAGS
07:35:12 pull-e2e-5-master-https pull-e2e-parallel-5 0.0.0.0/0  tcp:443          pull-e2e-5-master
07:35:15 Created [https://www.googleapis.com/compute/v1/projects/kubernetes-jenkins-pull/zones/us-central1-f/disks/pull-e2e-5-kube-system-kube-registry].
07:35:15 NAME                                 ZONE          SIZE_GB TYPE        STATUS
07:35:15 pull-e2e-5-kube-system-kube-registry us-central1-f 200     pd-standard READY
07:35:19 ERROR: (gcloud.compute.addresses.create) Some requests did not succeed:
07:35:19  - Quota 'STATIC_ADDRESSES' exceeded.  Limit: 25.0
07:35:19 
07:35:19 2015/09/28 07:35:19 e2e.go:307: Error running up: exit status 1
07:35:19 2015/09/28 07:35:19 e2e.go:303: Step 'up' finished in 39.755732144s
07:35:19 2015/09/28 07:35:19 e2e.go:141: Error starting e2e cluster. Aborting.
```
@cjcullen, @ixdy primary fyi

**Comment by derekwaynecarr** [comment]
Can you update ns_controller to ensure you delete the ingress resource when the namespace that contains it is purged so we do not chase future e2e test failures in this area?

See:
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/namespace/namespace_controller.go#L195

**Comment by bprashanth** [comment]
> Can you update ns_controller to ensure you delete the ingress resource when the namespace that contains it is purged so we do not chase future e2e test failures in this area?

@derekwaynecarr done. However I haven't added a controller-manager flag. will do that in a follow up pr as I really don't want to mix controller manager changes in here (also more flags to controller manager is yucky, so I might come up with a better plan)

File: pkg/controller/namespace/namespace_controller.go [hunk 16]
```diff
@@ -478,3 +482,17 @@ func deleteDeployments(expClient client.ExperimentalInterface, ns string) error
 	}
 	return nil
 }
+
+func deleteIngress(expClient client.ExperimentalInterface, ns string) error {
+	items, err := expClient.Ingress(ns).List(labels.Everything(), fields.Everything())
+	if err != nil {
+		return err
```
Comment (derekwaynecarr) [inline pkg/controller/namespace/namespace_controller.go]: Since we don't yet have a flag, do not return an error if you get a 404

Comment (bprashanth) [inline pkg/controller/namespace/namespace_controller.go]: Hmm, it's either all or nothing, i mean the apiserver either has all experimental resources (exp=true) or doesn't. We're only in trouble if the controller manager is started with one of the exp flags and the apiserver isn't, in which case the autoscaler itself will fail. We have disjoint sets of flags that denote the same thing.

Comment (derekwaynecarr) [inline pkg/controller/namespace/namespace_controller.go]: Sorry, this is fine.   I couldn't tell easily on phone if this was in the
experimental if block.  Thanks!

**Comment by bprashanth** [comment]
@k8s-bot test this please

**Comment by bprashanth** [comment]
Tim, trap is green
